python/mxnet/optimizer.py

```diff
@@ -771,16 +772,35 @@ class Updater(object):
     def __init__(self, optimizer):
         self.optimizer = optimizer
         self.states = {}
+        self.states_synced = {}
 
     def __call__(self, index, grad, weight):
         """Updates weight given gradient and index."""
         if index not in self.states:
             self.states[index] = self.optimizer.create_state(index, weight)
+            self.states_synced[index] = True
+        elif not self.states_synced[index]:
+            self.states[index] = \
+                self.sync_state_context(self.states[index], weight.context)
+            self.states_synced[index] = True
         self.optimizer.update(index, weight, grad, self.states[index])
 
+    def sync_state_context(self, state, context):
+        if isinstance(state, NDArray):
+            return state.as_in_context(context)
+        elif isinstance(state, tuple):
```

Comment:
Use isinstance(state, (tuple, list))

Also this can be done recursively, i.e.
return (self.sync_state_context(self, i, context) for i in state)